### PR TITLE
[dem] ProcessInfo is passed as const in many places

### PR DESCRIPTION
**Pull request by maceligueta**
**Description**
ProcessInfo is passed as `const`  as required by #7381.

**Changelog**
- `const` added everywhere
- Refactoring of the code due to the presence of changes in the ProcessInfo in inner functions of the element and constitutive law. Design was not ok and needed a refactor.

### CODE REVIEW

File: applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp
Comment (philbucher) [inline applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp]: I would use the parallel utilities and the atomic utiltites here

Comment (maceligueta) [reply applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp]: I just pushed this.

**Comment by maceligueta** [comment]
Hi @philbucher  I am getting an error with the atomic operation:
`D:\a\Kratos\Kratos\kratos\utilities/atomic_utilities.h(87,13): error C3005: 'write': unexpected token encountered on OpenMP 'atomic' directive (compiling source file D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\Custom\cotire\KratosDEMCore_CXX_unity_0_42.cxx) [D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\KratosDEMCore_unity.vcxproj]
D:\a\Kratos\Kratos\applications\DEMApplication\custom_strategies\strategies\explicit_solver_continuum.cpp(202): message : see reference to function template instantiation 'void Kratos::AtomicAssign<bool>(TDataType &,const TDataType &)' being compiled [D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\KratosDEMCore_unity.vcxproj]
          with
          [
              TDataType=bool
          ] (compiling source file D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\Custom\cotire\KratosDEMCore_CXX_unity_0_42.cxx)`

Maybe `critical` would work here. I was actually switching to `critical` when I received your comment. Because it seems like the `atomic write` operation with `bool` is not accepted, no?

**Comment by philbucher** [comment]
> Hi @philbucher I am getting an error with the atomic operation:
> `D:\a\Kratos\Kratos\kratos\utilities/atomic_utilities.h(87,13): error C3005: 'write': unexpected token encountered on OpenMP 'atomic' directive (compiling source file D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\Custom\cotire\KratosDEMCore_CXX_unity_0_42.cxx) [D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\KratosDEMCore_unity.vcxproj] D:\a\Kratos\Kratos\applications\DEMApplication\custom_strategies\strategies\explicit_solver_continuum.cpp(202): message : see reference to function template instantiation 'void Kratos::AtomicAssign<bool>(TDataType &,const TDataType &)' being compiled [D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\KratosDEMCore_unity.vcxproj] with [ TDataType=bool ] (compiling source file D:\a\Kratos\Kratos\build\Custom\applications\DEMApplication\Custom\cotire\KratosDEMCore_CXX_unity_0_42.cxx)`
> 
> Maybe `critical` would work here. I was actually switching to `critical` when I received your comment. Because it seems like the `atomic write` operation with `bool` is not accepted, no?

hm good question, unfortunately I am not the expert in this, but @ricc

**Comment by philbucher** [comment]
@RiccardoRossi can you answer this?

@maceligueta in this I recommend using the [`LockObject`](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/lock_object.h)
We are planning to also implement this to be compatible with C++11 based parallelsim

**Comment by maceligueta** [comment]
I just came across @RiccardoRossi and he knows that `#pragma atomic write` does not compile on Windows (requires OpenMP 3.0), so maybe this should not be in Kratos for the moment:
https://github.com/KratosMultiphysics/Kratos/blob/a5189a5ed4d9c1d67cd70e0bcb169bcf80e90151/kratos/utilities/atomic_utilities.h#L86
@RiccardoRossi  also bets for `locks `better than `critical`. I will implement it soon and push if nobody says anything against it.

**Comment by philbucher** [comment]
hm maybe we can make the `write` to only be added non-win compilation (@RiccardoRossi)

sorry for the noise in your PR @maceligueta

**Comment by maceligueta** [comment]
I think that the Codacy warning is a false positive... Apart from that I think I can believe with that `int `instead of a `bool`.

File: applications/DEMApplication/custom_elements/spheric_particle.h
Comment (philbucher) [inline applications/DEMApplication/custom_elements/spheric_particle.h]: I thin this can now be removed as we have the baseclass like this

**Review by philbucher — DISMISSED** [review]
looks good to me, thanks @maceligueta 

up to you to wait for approval of @KratosMultiphysics/dem

**Comment by RiccardoRossi** [comment]
as a comment, we coukd maybe use the "pragma single" in win (not 100% sure
it works) and the atomic where it is supported

File: applications/DEMApplication/custom_elements/spheric_particle.h
Comment (maceligueta) [reply applications/DEMApplication/custom_elements/spheric_particle.h]: Do you mean the line
`using DiscreteElement::Initialize;`  ?

Comment (philbucher) [reply applications/DEMApplication/custom_elements/spheric_particle.h]: yes

Comment (maceligueta) [reply applications/DEMApplication/custom_elements/spheric_particle.h]: I just removed it and that made me see that one file should not be there, so it was removed too. Thanks for your comments @philbucher .

**Review by philbucher — APPROVED** [review]

**Pull request merged**
